testsuite/features/step_definitions/common_steps.rb

```diff
@@ -70,18 +70,24 @@
   step %(I wait at most #{DEFAULT_TIMEOUT} seconds until event "#{event}" is completed)
 end
 
-When(/^I wait at most (\d+) seconds until event "([^"]*)" is completed$/) do |final_timeout, event|
+When(/^I wait at most (\d+) seconds until event "([^"]*)" is completed$/) do |timeout, event|
+  starting = Time.now
   # The code below is not perfect because there might be other events with the
   # same name in the events history - however, that's the best we have so far.
   steps %(
     When I follow "Events"
     And I follow "Pending"
-    And I wait until I do not see "#{event}" text, refreshing the page
+    And I wait at most #{timeout} seconds until I do not see "#{event}" text, refreshing the page
     And I follow "History"
     And I wait until I see "System History" text
     And I wait until I see "#{event}" text, refreshing the page
+  )
+  ending = Time.now
+  timeout -= ending - starting
+  timeout = 3 if timeout < 3
```

Comment:
Do we really need to use a different timeout value for the next step?
I'd use the timeout value coming from the step definition parameter in both steps without taking into account this difference of seconds "spent" in the first block. The code would be much easier and the advantage of having a bit smaller timeout is in my opinion none.

Reply:
It's a matter of semantics of this step.

We want to wait "at most" XX seconds in this step, but if we do like you say we will in fact spend at most XX seconds in first timeout and agains at most XX seconds in second step. For a total wait that can be up to 2 * XX seconds.

Now, if we accept that the step does not do exactly like the English sentence pretends it does, we can do like you want. It will indeed produce simpler code.

Alternatively, we could pass 2 parameters to the event, one for the time spent in "pending", and the other one for the time spent in running the event.

Reply:
Ok, I understand now your concern :)
Might be that we want to achieve too much in the same wrapper step definition? Might be that we need to write a bit more in our scenarios to clarify it? I'm asking my self, not sure, but when we use that much numbers of steps inside a step, having even some extra logic between them, maybe is because we are trying to wrap to much.